File: algopt/rebalancer/examples/website/cookbook/colocation/basic_colocation.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Basic Service Colocation for Microservices
 *
 * This example demonstrates how to colocate related microservices together
 * for performance benefits while maintaining fault tolerance through diversity.
 *
 * Problem: 60 service instances (20 frontend, 20 API, 20 cache) need to be
 * deployed across 20 hosts. Related services benefit from colocation (low
 * latency) but each service needs diversity across hosts for fault tolerance.
 *
 * Goal: Maximize colocation of service pairs while spreading each service type
 * across multiple hosts.
 */

// solution_start
#include "algopt/rebalancer/interface/ProblemSolver.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/ProblemSpecs_types.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/SolverSpecs_types.h"

#include <folly/container/irange.h>
#include <folly/executors/CPUThreadPoolExecutor.h>
#include <folly/init/Init.h>

#include <iostream>
#include <map>
#include <memory>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

void colocateMicroservices() {
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "microservice-colocation", "production");

  solver.setObjectName("instance");
  solver.setContainerName("host");

  // Service configuration
  struct ServiceConfig {
    int count;
    double cpu;
    double memory;
  };

  std::map<std::string, ServiceConfig> services = {
      {"frontend", {.count = 20, .cpu = 2.0, .memory = 4.0}},
      {"api", {.count = 20, .cpu = 4.0, .memory = 8.0}},
      {"cache", {.count = 20, .cpu = 1.0, .memory = 16.0}},
  };

  // Initial assignment (spread across all hosts)
  std::map<std::string, std::vector<std::string>> assignment;
  for (const auto i : folly::irange(20)) {
    assignment["host" + std::to_string(i)] = {};
  }

  int instance_id = 0;
  for (const auto& [service_name, config] : services) {
    for (const auto i : folly::irange(config.count)) {
      std::string instance_name =
          service_name + "_instance_" + std::to_string(i);
      int host_idx = instance_id % 20;
      assignment["host" + std::to_string(host_idx)].push_back(instance_name);
      instance_id++;
    }
  }

  solver.setAssignment(assignment);

  // Define service_type partition
  std::map<std::string, std::vector<std::string>> service_type_partition;
  for (const auto& [service_name, config] : services) {
    std::vector<std::string> instances;
    instances.reserve(config.count);
    for (const auto i : folly::irange(config.count)) {
      instances.push_back(service_name + "_instance_" + std::to_string(i));
    }
    service_type_partition[service_name] = instances;
  }
  solver.addPartition("service_type", std::move(service_type_partition));

  // Define service_pair partition
  std::map<std::string, std::vector<std::string>> service_pair_partition;

  // Frontend-API pairs
  for (const auto i : folly::irange(20)) {
    std::string pair_name = "frontend_api_pair_" + std::to_string(i);
    service_pair_partition[pair_name] = {
        "frontend_instance_" + std::to_string(i),
        "api_instance_" + std::to_string(i),
    };
  }

  // API-Cache pairs
  for (const auto i : folly::irange(20)) {
    std::string pair_name = "api_cache_pair_" + std::to_string(i);
    service_pair_partition[pair_name] = {
        "api_instance_" + std::to_string(i),
        "cache_instance_" + std::to_string(i),
    };
  }

  solver.addPartition("service_pair", std::move(service_pair_partition));

  // Instance resource usage
  std::map<std::string, double> cpu_usage;
  std::map<std::string, double> memory_usage;
  for (const auto& [service_name, config] : services) {
    for (const auto i : folly::irange(config.count)) {
      std::string instance_name =
          service_name + "_instance_" + std::to_string(i);
      cpu_usage[instance_name] = config.cpu;
      memory_usage[instance_name] = config.memory;
    }
  }
  solver.addObjectDimension("cpu_usage", std::move(cpu_usage));
  solver.addObjectDimension("memory_usage", std::move(memory_usage));

  // Host capacities
  std::map<std::string, double> host_cpu_capacity;
  std::map<std::string, double> host_memory_capacity;
  for (const auto i : folly::irange(20)) {
    std::string host = "host" + std::to_string(i);
    host_cpu_capacity[host] = 32.0;
    host_memory_capacity[host] = 128.0;
  }
  solver.addContainerDimension("cpu_capacity", host_cpu_capacity);
  solver.addContainerDimension("memory_capacity", host_memory_capacity);

  // CONSTRAINT 1: CPU capacity
  CapacitySpec cpu_cap;
  cpu_cap.name() = "cpu-capacity";
  cpu_cap.scope() = "host";
  cpu_cap.dimension() = "cpu_usage";
  solver.addConstraint(std::move(cpu_cap));

  // CONSTRAINT 2: Memory capacity
  CapacitySpec mem_cap;
  mem_cap.name() = "memory-capacity";
  mem_cap.scope() = "host";
  mem_cap.dimension() = "memory_usage";
  solver.addConstraint(std::move(mem_cap));

  // CONSTRAINT 3: Max instances per service per host
  GroupCountSpec max_per_host;
  max_per_host.name() = "max-instances-per-service-per-host";
  max_per_host.scope() = "host";
  max_per_host.partitionName() = "service_type";
  max_per_host.bound() = GroupCountSpecBound::MAX;
  Limit limit;
  limit.type() = LimitType::ABSOLUTE;
  limit.globalLimit() = 5;
  max_per_host.limit() = std::move(limit);
  solver.addConstraint(std::move(max_per_host));

  // GOAL 1: Colocate service pairs
  ColocateGroupsSpec colocate;
  colocate.name() = "colocate-service-pairs";
  colocate.scope() = "host";
  colocate.partitionName() = "service_pair";
  colocate.bound() = ColocateGroupsSpecBound::MAX;
  Limit coloc_limit;
  coloc_limit.type() = LimitType::ABSOLUTE;
  coloc_limit.globalLimit() = 1;
  colocate.limits() = std::move(coloc_limit);
  solver.addGoal(std::move(colocate), 5.0);

  // GOAL 2: Spread services
  ColocateGroupsSpec spread;
  spread.name() = "spread-services";
  spread.scope() = "host";
  spread.partitionName() = "service_type";
  spread.bound() = ColocateGroupsSpecBound::MIN;
  Limit spread_limit;
  spread_limit.type() = LimitType::ABSOLUTE;
  spread_limit.globalLimit() = 10;
  spread.limits() = std::move(spread_limit);
  solver.addGoal(std::move(spread), 3.0);

  // GOAL 3: Balance CPU
  BalanceSpec balance_cpu;
  balance_cpu.name() = "balance-cpu";
  balance_cpu.scope() = "host";
  balance_cpu.dimension() = "cpu_usage";
  balance_cpu.formula() = BalanceSpecFormula::LEGACY;
  balance_cpu.fixAverageToInitial() = true;
  solver.addGoal(std::move(balance_cpu), 1.0);

  // GOAL 4: Balance memory
  BalanceSpec balance_mem;
  balance_mem.name() = "balance-memory";
  balance_mem.scope() = "host";
  balance_mem.dimension() = "memory_usage";
  balance_mem.formula() = BalanceSpecFormula::LEGACY;
  balance_mem.fixAverageToInitial() = true;
  solver.addGoal(std::move(balance_mem), 1.0);

  // GOAL 5: Minimize movement
  MinimizeMovementSpec min_move;
  min_move.name() = "minimize-movement";
  min_move.scope() = "host";
  min_move.dimension() = "cpu_usage";
  solver.addGoal(std::move(min_move), 0.2);

  // Use Local Search solver
  LocalSearchSolverSpec ls_solver;
  ls_solver.moveTypeList()->push_back(
      ProblemSolver::makeMoveTypeSpec(SingleMoveTypeSpec()));
  ls_solver.solveTime() = 60000;
  solver.addSolver(ls_solver);

  // Solve
  auto solution = solver.solve();

  std::cout << "AssignmentSolution found in "
            << *solution.problemProfile()->solveSec() << " seconds\n";
  std::cout << "Objective value: " << *solution.finalObjective()->value()
            << "\n";

  int move_count = 0;
  if (!solution.solverSummaries()->empty()) {
    auto& first_solver = solution.solverSummaries()->at(0);
    if (first_solver.moveStats()) {
      move_count = *first_solver.moveStats()->numMoves();
    }
  }
  std::cout << "Instances moved: " << move_count << "\n";
}

int main(int argc, char* argv[]) {
  const folly::Init init(&argc, &argv);
  colocateMicroservices();
  return 0;
}
// solution_end

File: algopt/rebalancer/examples/website/cookbook/colocation/basic_colocation.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Basic Service Colocation for Microservices

This example demonstrates how to colocate related microservices together
for performance benefits while maintaining fault tolerance through diversity.

Problem: 60 service instances (20 frontend, 20 API, 20 cache) need to be
deployed across 20 hosts. Related services benefit from colocation (low latency)
but each service needs diversity across hosts for fault tolerance.

Goal: Maximize colocation of service pairs while spreading each service type
across multiple hosts.
"""

# solution_start
# pyre-strict
from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import (
    ConstraintSpecs,
    GoalSpecs,
    SolverSpecs,
)
from rebalancer.interface.thrift.v2.ProblemSpecs.thrift_types import (
    BalanceSpec,
    BalanceSpecFormula,
    CapacitySpec,
    ColocateGroupsSpec,
    ColocateGroupsSpecBound,
    GroupCountSpec,
    GroupCountSpecBound,
    Limit,
    LimitType,
    MinimizeMovementSpec,
)
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    LocalSearchSolverSpec,
)


def colocate_microservices():
    """Colocate related microservices for performance."""
    solver = ProblemSolver(
        service_name="microservice-colocation", service_scope="production"
    )

    solver.setObjectName("instance")
    solver.setContainerName("host")

    # Service configuration
    services = {
        "frontend": {"count": 20, "cpu": 2.0, "memory": 4.0},
        "api": {"count": 20, "cpu": 4.0, "memory": 8.0},
        "cache": {"count": 20, "cpu": 1.0, "memory": 16.0},
    }

    # Initial spread assignment (all services evenly distributed)
    assignment = {f"host{i}": [] for i in range(20)}

    instance_id = 0
    for service_name, config in services.items():
        for i in range(config["count"]):
            instance_name = f"{service_name}_instance_{i}"
            # Initially spread across all hosts
            host_idx = instance_id % 20
            assignment[f"host{host_idx}"].append(instance_name)
            instance_id += 1

    solver.setAssignment(assignment)

    # Define service_type partition (group by service)
    service_type_partition = {}
    for service_name, config in services.items():
        service_type_partition[service_name] = [
            f"{service_name}_instance_{i}" for i in range(config["count"])
        ]
    solver.addPartition("service_type", service_type_partition)

    # Define colocation pairs partition
    # Each group represents instances that should be colocated
    service_pair_partition = {}

    # Frontend-API pairs (instance 0 with instance 0, etc.)
    for i in range(20):
        pair_name = f"frontend_api_pair_{i}"
        service_pair_partition[pair_name] = [
            f"frontend_instance_{i}",
            f"api_instance_{i}",
        ]

    # API-Cache pairs
    for i in range(20):
        pair_name = f"api_cache_pair_{i}"
        service_pair_partition[pair_name] = [
            f"api_instance_{i}",
            f"cache_instance_{i}",
        ]

    solver.addPartition("service_pair", service_pair_partition)

    # Instance resource usage
    cpu_usage = {}
    memory_usage = {}
    for service_name, config in services.items():
        for i in range(config["count"]):
            instance_name = f"{service_name}_instance_{i}"
            cpu_usage[instance_name] = config["cpu"]
            memory_usage[instance_name] = config["memory"]

    solver.addObjectDimension("cpu_usage", cpu_usage)
    solver.addObjectDimension("memory_usage", memory_usage)

    # Host capacities (uniform)
    host_cpu_capacity = {f"host{i}": 32.0 for i in range(20)}  # 32 cores
    host_memory_capacity = {f"host{i}": 128.0 for i in range(20)}  # 128 GB
    solver.addContainerDimension("cpu_capacity", host_cpu_capacity)
    solver.addContainerDimension("memory_capacity", host_memory_capacity)

    # CONSTRAINT 1: CPU capacity
    solver.addConstraint(
        ConstraintSpecs(
            capacitySpec=CapacitySpec(
                name="cpu-capacity",
                scope="host",
                dimension="cpu_usage",
            )
        )
    )

    # CONSTRAINT 2: Memory capacity
    solver.addConstraint(
        ConstraintSpecs(
            capacitySpec=CapacitySpec(
                name="memory-capacity",
                scope="host",
                dimension="memory_usage",
            )
        )
    )

    # CONSTRAINT 3: Max 5 instances per service per host (diversity)
    solver.addConstraint(
        ConstraintSpecs(
            groupCountSpec=GroupCountSpec(
                name="max-instances-per-service-per-host",
                scope="host",
                partitionName="service_type",
                bound=GroupCountSpecBound.MAX,
                limit=Limit(type=LimitType.ABSOLUTE, globalLimit=5),
            )
        )
    )

    # GOAL 1: Colocate service pairs (MAX = keep pairs on same host)
    solver.addGoal(
        GoalSpecs(
            colocateGroupsSpec=ColocateGroupsSpec(
                name="colocate-service-pairs",
                scope="host",
                partitionName="service_pair",
                bound=ColocateGroupsSpecBound.MAX,
                limits=Limit(type=LimitType.ABSOLUTE, globalLimit=1),
            )
        ),
        weight=5.0,  # High weight - colocation is important for performance
    )

    # GOAL 2: Spread each service across hosts (MIN = spread out)
    solver.addGoal(
        GoalSpecs(
            colocateGroupsSpec=ColocateGroupsSpec(
                name="spread-services",
                scope="host",
                partitionName="service_type",
                bound=ColocateGroupsSpecBound.MIN,
                limits=Limit(type=LimitType.ABSOLUTE, globalLimit=10),
            )
        ),
        weight=3.0,  # Medium weight - diversity for fault tolerance
    )

    # GOAL 3: Balance CPU across hosts
    solver.addGoal(
        GoalSpecs(
            balanceSpec=BalanceSpec(
                name="balance-cpu",
                scope="host",
                dimension="cpu_usage",
                formula=BalanceSpecFormula.LEGACY,
                fixAverageToInitial=True,
            )
        ),
        weight=1.0,
    )

    # GOAL 4: Balance memory across hosts
    solver.addGoal(
        GoalSpecs(
            balanceSpec=BalanceSpec(
                name="balance-memory",
                scope="host",
                dimension="memory_usage",
                formula=BalanceSpecFormula.LEGACY,
                fixAverageToInitial=True,
            )
        ),
        weight=1.0,
    )

    # GOAL 5: Minimize movement
    solver.addGoal(
        GoalSpecs(
            minimizeMovementSpec=MinimizeMovementSpec(
                name="minimize-movement",
                scope="host",
                dimension="cpu_usage",
            )
        ),
        weight=0.2,
    )

    # Use Local Search solver
    solver.addSolver(
        SolverSpecs(localSearchSolverSpec=LocalSearchSolverSpec(timeLimitMs=60000))
    )

    # Solve
    solution = solver.solve()

    # Analyze solution
    print(f"Solution found in {solution.profile.solveTime}ms")
    print(f"Objective value: {solution.objectiveValue}")
    print(f"Instances moved: {solution.profile.moveCount}")

    # Analyze colocation results
    print("\nColocation analysis:")
    analyze_colocation(solution.assignment, service_pair_partition)

    # Analyze service spread
    print("\nService spread analysis:")
    analyze_service_spread(solution.assignment, service_type_partition)

    return solution


def analyze_colocation(assignment, service_pair_partition):
    """Analyze how well service pairs are colocated."""
    colocated_pairs = 0
    split_pairs = 0

    for instances in service_pair_partition.values():
        # Find which hosts have these instances
        hosts_with_pair = set()
        for host, host_instances in assignment.items():
            if any(instance in host_instances for instance in instances):
                hosts_with_pair.add(host)

        if len(hosts_with_pair) == 1:
            colocated_pairs += 1
        else:
            split_pairs += 1

    total_pairs = len(service_pair_partition)
    colocation_pct = (colocated_pairs / total_pairs) * 100

    print(f"  Colocated pairs: {colocated_pairs}/{total_pairs} ({colocation_pct:.1f}%)")
    print(f"  Split pairs: {split_pairs}/{total_pairs}")


def analyze_service_spread(assignment, service_type_partition):
    """Analyze service diversity across hosts."""
    for service_name, instances in service_type_partition.items():
        # Count hosts with this service
        hosts_with_service = set()
        for host, host_instances in assignment.items():
            if any(instance in instances for instance in host_instances):
                hosts_with_service.add(host)

        instance_count = len(instances)
        host_count = len(hosts_with_service)

        print(f"  {service_name}: {instance_count} instances on {host_count} hosts")


if __name__ == "__main__":
    colocate_microservices()
# solution_end
